File: features/auth/fastify/files/src/modules/auth/session.ts
import type { User } from '@prisma/client';
import type { FastifyInstance, FastifyRequest } from 'fastify';
import { createHash, randomUUID } from 'node:crypto';

export const REFRESH_TTL_SECONDS = 7 * 24 * 60 * 60;

export const ROLE_MODEL = {
  roles: ['admin', 'user'] as const,
  permissions_by_role: {
    admin: ['*'],
    user: ['*:read.*'],
  },
};

export function hashRefreshToken(token: string): string {
  return createHash('sha256').update(token).digest('hex');
}

export function permissionsForRole(role: string): string[] {
  const perms = ROLE_MODEL.permissions_by_role[role as keyof typeof ROLE_MODEL.permissions_by_role];
  return perms ?? [];
}

export interface AuthTokenPayload {
  sub: string;
  role: string;
  sid: string;
  email: string;
  name?: string;
  permissions: string[];
  token_type: 'access' | 'refresh';
  jti: string;
}

export function signTokens(
  fastify: {
    jwt: {
      sign: (payload: Record<string, unknown>, options?: { expiresIn?: string }) => string;
    };
  },
  payload: Omit<AuthTokenPayload, 'token_type' | 'jti'>,
) {
  const accessJti = randomUUID();
  const refreshJti = randomUUID();
  const token = fastify.jwt.sign(
    { ...payload, token_type: 'access', jti: accessJti },
    { expiresIn: '15m' },
  );
  const refresh_token = fastify.jwt.sign(
    { ...payload, token_type: 'refresh', jti: refreshJti },
    { expiresIn: '7d' },
  );
  return { token, refresh_token, access_jti: accessJti, refresh_jti: refreshJti };
}

export async function issueAuthSession(
  fastify: FastifyInstance,
  user: User,
  request: FastifyRequest,
): Promise<{
  user: Record<string, unknown>;
  token: string;
  access_token: string;
  refresh_token: string;
}> {
  const sessionId = randomUUID();
  const payload = {
    sub: user.id,
    sid: sessionId,
    role: user.role,
    email: user.email,
    name: user.name,
    permissions: [...permissionsForRole(user.role)],
  };
  const tokens = signTokens(fastify, payload);
  const expiresAt = new Date(Date.now() + REFRESH_TTL_SECONDS * 1000);

  await fastify.prisma.refreshToken.create({
    data: {
      user_id: user.id,
      session_id: sessionId,
      token_hash: hashRefreshToken(tokens.refresh_token),
      expires_at: expiresAt,
      ip_address: request.ip,
      user_agent: request.headers['user-agent'] ?? null,
    },
  });

  return {
    user: {
      id: user.id,
      email: user.email,
      name: user.name,
      role: user.role,
      last_login: user.last_login,
      created_at: user.created_at,
      updated_at: user.updated_at,
    },
    token: tokens.token,
    access_token: tokens.token,
    refresh_token: tokens.refresh_token,
  };
}

File: features/auth/fastapi/files/src/auth/_session.py
import hashlib
import os
import uuid
from datetime import UTC, datetime, timedelta
from typing import Any

import jwt
from sqlalchemy.ext.asyncio import AsyncSession

from ._models import RefreshToken, User

REFRESH_TTL_SECONDS = 7 * 24 * 60 * 60
ACCESS_TTL_SECONDS = 15 * 60
MFA_CHALLENGE_TTL_SECONDS = 5 * 60

ROLE_PERMISSIONS: dict[str, list[str]] = {
    "admin": ["*:*.*"],
    "user": ["*:read.*"],
}


def permissions_for_role(role: str) -> list[str]:
    return list(ROLE_PERMISSIONS.get(role, []))


def _jwt_secret() -> str:
    secret = os.getenv("JWT_SECRET")
    if not secret:
        raise RuntimeError("JWT_SECRET is required to issue tokens")
    return secret


def _jwt_algorithm() -> str:
    raw = os.getenv("JWT_ALGORITHMS", "HS256").split(",")
    return raw[0].strip() or "HS256"


def hash_refresh_token(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def _sign(payload: dict[str, Any], expires_in_seconds: int) -> str:
    now = datetime.now(UTC)
    full = {
        **payload,
        "iat": int(now.timestamp()),
        "exp": int((now + timedelta(seconds=expires_in_seconds)).timestamp()),
    }
    return jwt.encode(full, _jwt_secret(), algorithm=_jwt_algorithm())


def sign_mfa_challenge(user_id: str) -> str:
    return _sign({"sub": user_id, "stage": "mfa_pending"}, MFA_CHALLENGE_TTL_SECONDS)


def verify_mfa_challenge(token: str) -> dict[str, Any]:
    return jwt.decode(token, _jwt_secret(), algorithms=[_jwt_algorithm()])


def sign_tokens(payload: dict[str, Any]) -> dict[str, str]:
    access_jti = str(uuid.uuid4())
    refresh_jti = str(uuid.uuid4())
    access_token = _sign(
        {**payload, "token_type": "access", "jti": access_jti},
        ACCESS_TTL_SECONDS,
    )
    refresh_token = _sign(
        {**payload, "token_type": "refresh", "jti": refresh_jti},
        REFRESH_TTL_SECONDS,
    )
    return {
        "token": access_token,
        "access_token": access_token,
        "refresh_token": refresh_token,
        "access_jti": access_jti,
        "refresh_jti": refresh_jti,
    }


def _serialize_user(user: User) -> dict[str, Any]:
    return {
        "id": user.id,
        "email": user.email,
        "name": user.name,
        "role": user.role,
        "last_login": user.last_login,
        "created_at": user.created_at,
        "updated_at": user.updated_at,
    }


async def issue_auth_session(
    session: AsyncSession,
    user: User,
    ip_address: str | None,
    user_agent: str | None,
) -> dict[str, Any]:
    session_id = str(uuid.uuid4())
    payload = {
        "sub": user.id,
        "sid": session_id,
        "role": user.role,
        "email": user.email,
        "name": user.name,
        "permissions": permissions_for_role(str(user.role)),
    }
    tokens = sign_tokens(payload)
    expires_at = datetime.now(UTC) + timedelta(seconds=REFRESH_TTL_SECONDS)
    refresh_row = RefreshToken(
        user_id=user.id,
        session_id=session_id,
        token_hash=hash_refresh_token(tokens["refresh_token"]),
        expires_at=expires_at,
        ip_address=ip_address,
        user_agent=user_agent,
    )
    session.add(refresh_row)
    await session.flush()
    return {
        "user": _serialize_user(user),
        "token": tokens["token"],
        "access_token": tokens["access_token"],
        "refresh_token": tokens["refresh_token"],
    }
